### DISCUSSION TITLE
parse decimal as string

### DISCUSSION THREAD

**Discussion by Golden-1** [simdjson/simdjson#2015] [Q&A]
can i parse decimal to string when parse a json? for example:for example {"name": "Jason", "total_count": 590990000909011119999},directly parse 590990000909011119999 to "590990000909011119999"?

**Comment by lemire** [comment]
Yes. You can get direct access to the raw data by calling `raw_json_token()`.

```C++
simdjson::ondemand::parser parser;
simdjson::padded_string docdata =  R"({"value":12321323213213213213213213213211223})"_padded;
simdjson::ondemand::document doc = parser.iterate(docdata);
simdjson::ondemand::object obj = doc.get_object();
std::string_view token = obj["value"].raw_json_token();
```

This will give you a string pointing at `12321323213213213213213213213211223` which you may process as you see fit.

See https://github.com/simdjson/simdjson/blob/master/doc/basics.md#general-direct-access-to-the-raw-json-string

**Reply by Golden-1** [reply]
get，thanks a lot；btw，is the parsed result read only？are there any report about read performance？

**Reply by lemire** [reply]
> is the parsed result read only？are there any report about read performance？

A `raw_json_token()` give you direct access to the input string so that in your example `{"name": "Jason", "total_count": 590990000909011119999}`, you directly parse 590990000909011119999 to string "`590990000909011119999`"?

If you need it in, say, a decimal type, then you need to parse it using an appropriate library. [There is no decimal type in C++](https://stackoverflow.com/questions/14096026/c-decimal-data-types).

Once decimal types are part of the C++ standard, we will add support for it in simdjson. For the time being, we prefer to leave it up to our users.

**Reply by Golden-1** [reply]
I have got the last question；I have two new questions：1. if i receive a decimal type from client，and parse it to string through raw_json_token，will it lose precision? 2. after I parse json to dom，can i change some value in it？

**Reply by lemire** [reply]
> 1. if i receive a decimal type from client，and parse it to string through raw_json_token，will it lose precision?

I am not sure what `receive a decimal type from client` means because there is no such thing in JSON.

If you convert the element to floating-point number, through `get_double()`, then you get the nearest 64-bit floating-point approximation. This is what is expected in JSON. Please refer to the [JSON specification](https://www.tbray.org/ongoing/When/201x/2017/12/14/rfc8259.html):


> **This specification allows implementations to set limits on the range and precision of numbers accepted. Since software that implements IEEE 754 binary64 (double precision) numbers [[IEEE754]](https://xml2rfc.tools.ietf.org/cgi-bin/cat.cgi/rfc8259.html?input=a76e3edb9610a98045f4a1729262df1867b28097abc973cd69fdc1-1513534632#IEEE754) is generally available and widely used, good interoperability can be achieved by implementations that expect no more precision or range than these provide, in the sense that implementations will approximate JSON numbers within the expected precision. A JSON number such as 1E400 or 3.141592653589793238462643383279 may indicate potential interoperability problems, since it suggests that the software that created it expects receiving software to have greater capabilities for numeric magnitude and precision than is widely available.**

> **Note that when such software is used, numbers that are integers and are in the range [-(2**53)+1, (2**53)-1] are interoperable in the sense that implementations will agree exactly on their numeric values.**


Please recall that there is no decimal type in C++. 

If you call the `raw_json_token()` method, then you get access to a string (actually a `std::string_view` instance) which corresponds exactly to what appears in the JSON document. In fact, it points at the JSON document. So there is no loss possible.

If you have some implementation of a decimal type (and it will be assuredly not part of standard C++), then you can parse the result of `raw_json_token()`, but that's your responsibility, it is not part of the simdjson library.

> 2. after I parse json to dom，can i change some value in it？

The `raw_json_token()` method will return a `std::string_view` instance. A `std::string_view` instance is always immutable in the sense that you cannot change the underlying content. You are supposed to process the content for your own needs.

**Reply by Golden-1** [reply]
get，thank you very much
